File: server/services/smsService.js
/**
 * SMS Service for Node.js Backend
 * Using SMSONLINEGH API for SMS delivery
 */

const axios = require('axios');

const SMS_API_KEY = process.env.SMS_API_KEY;
const SMS_SENDER_ID = process.env.SMS_SENDER_ID;
const SMS_HOST = process.env.SMS_HOST || 'api.smsonlinegh.com';

// Use the public Render SMS gateway by default in production.
// If Vercel has SMS_GATEWAY_URL configured, it will override this.
const SMS_GATEWAY_URL = process.env.SMS_GATEWAY_URL || 'https://reektickets.onrender.com';
const USE_GATEWAY = process.env.USE_SMS_GATEWAY !== 'false';

console.log('[SMS] Service initialized with:');
console.log('[SMS] - API_HOST:', SMS_HOST);
console.log('[SMS] - SENDER_ID:', SMS_SENDER_ID ? 'SET' : 'NOT SET');
console.log('[SMS] - API_KEY:', SMS_API_KEY ? `SET (${SMS_API_KEY.substring(0,8)}...)` : 'NOT SET');
console.log('[SMS] - GATEWAY_URL:', SMS_GATEWAY_URL);
console.log('[SMS] - USE_GATEWAY:', USE_GATEWAY);

function formatPhone(phone) {
  let cleanPhone = phone.replace(/\s+/g, '').replace(/^\+/, '');
  if (cleanPhone.startsWith('0')) {
    cleanPhone = '233' + cleanPhone.substring(1);
  } else if (!cleanPhone.startsWith('233')) {
    cleanPhone = '233' + cleanPhone;
  }
  return cleanPhone;
}

async function sendViaGateway(phone, message) {
  try {
    console.log(`[SMS] Trying gateway: ${SMS_GATEWAY_URL}`);
    const response = await axios.post(`${SMS_GATEWAY_URL}/send-sms`, {
      phone,
      message
    }, {
      timeout: 20000,
      headers: {
        'Content-Type': 'application/json'
      }
    });

    if (response.data && response.data.success) {
      console.log('[SMS] Gateway sent successfully');
      return response.data;
    }

    console.error('[SMS] Gateway returned non-success', response.data);
    return null;
  } catch (error) {
    console.error('[SMS] Gateway failed:', error.message);
    return null;
  }
}

async function sendViaSmsonlinegh(phone, message) {
  try {
    const cleanPhone = formatPhone(phone);
    
    // Check if API key is set
    if (!SMS_API_KEY) {
      console.error('[SMS] ERROR: SMS_API_KEY environment variable not set');
      return {
        success: false,
        status: 500,
        error: 'SMS_API_KEY not configured',
        message: 'SMS service not properly configured'
      };
    }

    const params = {
      apikey: SMS_API_KEY,
      sender: SMS_SENDER_ID || 'ReekTickets',
      message,
      recipients: cleanPhone
    };

    const url = `https://${SMS_HOST}/sms/send/?${new URLSearchParams(params).toString()}`;
    console.log(`[SMS] SMSONLINEGH API call to ${cleanPhone}`);
    console.log(`[SMS] API Host: ${SMS_HOST}`);
    console.log(`[SMS] Sender ID: ${SMS_SENDER_ID || 'ReekTickets'}`);

    const response = await axios.get(url, {
      timeout: 20000,
      headers: {
        'User-Agent': 'ReekTickets-SMS/1.0',
        'Accept': 'application/json'
      },
      validateStatus: () => true
    });

    console.log(`[SMS] SMSONLINEGH response status: ${response.status}`);
    console.log(`[SMS] SMSONLINEGH response data:`, JSON.stringify(response.data).substring(0, 200));

    if (response.status === 200) {
      return {
        success: true,
        status: 200,
        data: response.data,
        message: `SMS sent successfully to ${cleanPhone}`
      };
    }

    return {
      success: false,
      status: response.status,
      error: `SMSONLINEGH response ${response.status}`,
      message: 'Failed to send SMS via SMSONLINEGH'
    };
  } catch (error) {
    console.error('[SMS] SMSONLINEGH error:', error.message);
    let errorMsg = error.message;
    if (error.response) {
      errorMsg = `HTTP ${error.response.status}: ${error.response.statusText}`;
    }
    console.error('[SMS] Full error:', error);
    return {
      success: false,
      status: 500,
      error: errorMsg,
      message: `SMSONLINEGH API error: ${errorMsg}`
    };
  }
}

async function sendSMS(phone, message) {
  try {
    console.log(`[SMS] Sending to ${phone}: ${message.substring(0, 50)}...`);
    console.log(`[SMS] SMS_API_KEY present: ${!!SMS_API_KEY}`);
    console.log(`[SMS] SMS_SENDER_ID: ${SMS_SENDER_ID}`);
    console.log(`[SMS] SMS_HOST: ${SMS_HOST}`);
    console.log(`[SMS] USE_GATEWAY: ${USE_GATEWAY}`);

    if (USE_GATEWAY) {
      const gatewayResult = await sendViaGateway(phone, message);
      if (gatewayResult && gatewayResult.success) {
        return gatewayResult;
      }
    }

    // If API key is not configured, return a helpful error
    if (!SMS_API_KEY) {
      console.error('[SMS] ERROR: SMS_API_KEY is not configured!');
      return {
        success: false,
        status: 500,
        error: 'SMS_API_KEY not configured',
        message: 'SMS service configuration error - please contact support'
      };
    }

    return await sendViaSmsonlinegh(phone, message);
  } catch (error) {
    console.error('[SMS] Unexpected error:', error.message);
    return {
      success: false,
      status: 500,
      error: error.message,
      message: 'SMS sending failed'
    };
  }
}

async function sendOTP(phone, otp) {
  const message = `Your ReekTickets verification code is ${otp}. It expires in 10 minutes.`;
  console.log(`[OTP] Sending OTP to ${phone}: "${message}"`);
  const result = await sendSMS(phone, message);
  console.log(`[OTP] Result:`, JSON.stringify(result));
  return result;
}

async function sendTicketConfirmation(phone, ticketDetails) {
  const message = `Your ReekTickets ticket is confirmed! Event: ${ticketDetails.eventName}, Date: ${ticketDetails.date}, Code: ${ticketDetails.code}`;
  return await sendSMS(phone, message);
}

async function healthCheck() {
  try {
    const result = await sendSMS('0273476701', 'SMS service health check');
    return {
      success: true,
      message: 'SMS service is operational',
      testResult: result.success
    };
  } catch (error) {
    return {
      success: false,
      message: 'SMS service health check failed',
      error: error.message
    };
  }
}

module.exports = {
  sendSMS,
  sendOTP,
  sendTicketConfirmation,
  healthCheck
};

File: server/sms-service.py
#!/usr/bin/env python3
"""
Flask SMS Service using Zenoph SDK
Deploy this to Vercel or Railway and call it from Node.js
"""

import os
import sys
import json
import traceback
from flask import Flask, request, jsonify

# Add Zenoph SDK to path
ZENOPH_PATH = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, os.path.join(ZENOPH_PATH, '../zenoph.notify-2.23.10-python'))

try:
    from Zenoph.Notify.Request.AuthRequest import AuthRequest
    from Zenoph.Notify.Request.SMSRequest import SMSRequest
except ImportError as e:
    print(f"[SMS Service] Warning: Zenoph SDK not installed - {e}")

app = Flask(__name__)

API_KEY = os.getenv('SMS_API_KEY', 'c6f61e914257462812deaff55c412a213cbf61a6388761016a1b2263d347948b')
SENDER_ID = os.getenv('SMS_SENDER_ID', 'ReekTickets')
HOST = "api.smsonlinegh.com"


def send_sms_zenoph(api_key, sender_id, phone, message):
    """Send SMS using Zenoph SDK"""
    try:
        # Authenticate
        auth_request = AuthRequest()
        auth_request.setHost(HOST)
        auth_request.setAuthApiKey(api_key)
        auth_profile = auth_request.authenticate()
        
        # Send SMS
        sms_request = SMSRequest(auth_profile)
        sms_request.setHost(HOST)
        sms_request.setSender(sender_id)
        sms_request.setMessage(message)
        sms_request.addDestination(phone)
        response = sms_request.submit()
        
        http_status = response.getHttpStatusCode()
        data_fragment = response.getDataFragment()
        
        return {
            'success': http_status == 200,
            'status': http_status,
            'data': str(data_fragment) if data_fragment else '',
            'message': f'SMS sent successfully to {phone}' if http_status == 200 else f'SMS send failed with status {http_status}'
        }
    except Exception as e:
        return {
            'success': False,
            'error': str(e),
            'message': f'SMS sending error: {str(e)}',
            'traceback': traceback.format_exc()
        }


@app.route('/health', methods=['GET'])
def health():
    """Health check endpoint"""
    return jsonify({
        'status': 'ok',
        'service': 'SMS Gateway',
        'api_key': 'SET' if API_KEY else 'NOT SET',
        'sender_id': SENDER_ID
    })


@app.route('/send', methods=['POST'])
def send_sms():
    """Send SMS endpoint"""
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({'error': 'No JSON data provided'}), 400
        
        phone = data.get('phone')
        message = data.get('message')
        
        if not phone or not message:
            return jsonify({'error': 'Missing phone or message'}), 400
        
        print(f"[SMS Service] Sending to {phone}: {message[:50]}...")
        
        result = send_sms_zenoph(API_KEY, SENDER_ID, phone, message)
        
        if result['success']:
            return jsonify(result), 200
        else:
            return jsonify(result), 500
    except Exception as e:
        return jsonify({
            'error': str(e),
            'message': 'Server error',
            'traceback': traceback.format_exc()
        }), 500


@app.route('/send-otp', methods=['POST'])
def send_otp():
    """Send OTP endpoint"""
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({'error': 'No JSON data provided'}), 400
        
        phone = data.get('phone')
        otp = data.get('otp')
        
        if not phone or not otp:
            return jsonify({'error': 'Missing phone or otp'}), 400
        
        message = f"Your ReekTickets verification code is {otp}. It expires in 10 minutes."
        print(f"[SMS Service] Sending OTP to {phone}: {otp}")
        
        result = send_sms_zenoph(API_KEY, SENDER_ID, phone, message)
        
        if result['success']:
            return jsonify(result), 200
        else:
            return jsonify(result), 500
    except Exception as e:
        return jsonify({
            'error': str(e),
            'message': 'Server error'
        }), 500


if __name__ == '__main__':
    port = int(os.getenv('PORT', 5001))
    app.run(host='0.0.0.0', port=port, debug=False)
